### PR TITLE
[Dubb - epoll ] Support epoll

### PR DESCRIPTION
**Pull request by zonghaishang**
## What is the purpose of the change

support epoll for dubbo, refer: https://github.com/apache/incubator-dubbo/issues/2737

## Brief changelog

1. consumerConfig, providerConfig, dubbo.xsd
2. AbstractSupport, NettySupport, NettyServer, NettyClient

## Verifying this change

ci passed & benchmark test passed.

benchmark: https://github.com/zonghaishang/dubbo-benchmark/tree/epoll_benchmark

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/AbstractSupport.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    public AbstractSupport(URL url) {
+        if (url == null) {
+            throw new IllegalArgumentException("url == null");
+        }
+        this.url = url;
+        this.nThreads = url.getPositiveParameter(Constants.IO_THREADS_KEY, Constants.DEFAULT_IO_THREADS);
+        this.epoll = isLinux() && url.getParameter(Constants.EPOLL_ENABLE, false);
+    }
+    
+    public NioEventLoopGroup nioEventLoopGroup(ThreadFactory threadFactory) {
```
[the remark this answers is not in the document]
Comment (zonghaishang) [reply dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/AbstractSupport.java]: epoll作为基类封装起来问题不大

**Comment by zonghaishang** [comment]
针对：按照目前的方式应该是会的，通过url会从provider透传到consumer，从而影响到consumer的epoll。
实际上感觉不透传好一些吧？

epoll server支持 epoll 和 非epoll调用过来，透传给client， 如果client自动知道server支持epoll，针对dubbo场景多provider 自动用epoll效果更好一些

File: dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/AbstractSupport.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    protected int nThreads;
+    protected boolean epoll;
+
+    public AbstractSupport(URL url) {
+        if (url == null) {
+            throw new IllegalArgumentException("url == null");
+        }
+        this.url = url;
+        this.nThreads = url.getPositiveParameter(Constants.IO_THREADS_KEY, Constants.DEFAULT_IO_THREADS);
+        this.epoll = isLinux() && url.getParameter(Constants.EPOLL_ENABLE, false);
```
Comment (htynkn) [inline dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/AbstractSupport.java]: linux system doesn't mean it support epoll. we also need check major version.
Or we can use `io.netty.channel.epoll.Epoll`, it has a method `isAvailable`

Comment (zonghaishang) [reply dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/AbstractSupport.java]: nice, I will fix it

**Comment by beiwei30** [comment]
没仔细看，几个建议：

* epoll 打开与否目前只和 netty4 绑定，合理的做法是和 protocol 上的 transporter 绑定。或者在应用级别来控制(虽然现在只有 netty4 才支持)。也就是说没有必要在 provider, consumer, protocol 上都有是否支持 epoll 的判断
* epoll 只是对 linux 平台上 native transport 的支持而已，更合理的做法是把 `AbstractSupport` 和 `NettySupport` 合并成 `NativeTransportFactory`，用来根据配置和平台自动返回对应的 `SocketChannel` 和 `EventLoopGroup` 的实例，当然出错情况下自动 fallbak 回 nio

综上所述，看起来 Transporter 应该有一个 TransporterConfig，或者干脆在 app 级别上定义。如果想不清楚，暂时通过 system properties 指定也是可以的。Native Transport 的支持需要考虑全平台，而不是只有 linux 上的 epoll，做成 factory 模式，出错 fallback 到 nio

---

I have a couple of suggestions after take a glance at the implementation:

* Right now, epoll support is tightly coupled with transport `netty4`, which I think it is fine, but from the perspective of its configuration, I believe more reasonable place for this configuration is transporter, or in my own opinion this configuration for whether enable native transport or now is global configuration, and should be applied to application level, even though currently we will only have its support for linux on netty4.  It is unnecessary and unreasonable to have the individual configuration for enabling epoll on different levels like provider, consumer, and protocol.
* Right now the implementation focus on epoll support only, we should consider more. Besides epoll on linux, we should also support kqueue on Mac/BSD, see https://netty.io/wiki/native-transports.html for more details. What I suggest is to implement a `NativeTransportFactory` by combining and enhancing the code in `AbstractSupport` and `NettySupport` together to produce the appropriate instances for `SocketChannel` and `EventLoopGroup` according to the platforms and the configurations the current application is running on. Of course, this factory should fallback to return the nio instances if anything goes wrong.

In summary, we should either define 'native-transporter' on ApplicationConfig or TransporterConfig (not introduced yet). It is also reasonable to introduce it with a system property if we cannot make the decision now.  For native transport support, we should consider not only epoll on linux, but also others.

**Comment by zonghaishang** [comment]
> epoll 打开与否目前只和 netty4 绑定，合理的做法是和 protocol 上的 transporter 绑定。或者在应用级别来控制(虽然现在只有 netty4 才支持)。也就是说没有必要在 provider, consumer, protocol 上都有是否支持 epoll 的判断

目前client端没有protocol承载属性，这块改不了，除非提前合并http2特性的代码

**Comment by lexburner** [comment]
+1, 之前测试过 Netty 支持的 native epoll, benchmark 的时候感触挺深，使用 NioSocketChannel 最高 qps 5600 ，替换成 EpollSocketChannel 之后 6300 qps，提升显著。但局限也是有的，1. 不兼容低版本 netty 2. 只在部分 linux 平台支持。
同意 @beiwei30 的建议，不需要在 dubbo 层面暴露 epoll 开关。

**Comment by beiwei30** [comment]
any further update on this pull request @zonghaishang

**Comment by zonghaishang** [comment]
Recently dealing with work, this feature will be completed by this week.

**Comment by lexburner** [comment]
Hi
Here are some benchmark between NioEventLoop and EpollEventLoop, may be helpful to us.
I will complete the full benchmark with multi concurrency this week.

## NioEventLoop

### loop 1

--------------------------
Durations:       60.05s
Requests:        149128
Avg RT:          199.96ms
Max RT:          956.371ms
Min RT:          70.502ms
Error requests:  0
Valid requests:  149128
QPS:             2483.59
--------------------------

### loop 2

--------------------------
Durations:       60.10s
Requests:        167389
Avg RT:          178.01ms
Max RT:          869.573ms
Min RT:          56.505ms
Error requests:  0
Valid requests:  167389
QPS:             2785.27
--------------------------

### loop3

--------------------------
Durations:       60.08s
Requests:        182469
Avg RT:          163.76ms
Max RT:          1596.385ms
Min RT:          56.073ms
Error requests:  0
Valid requests:  182469
QPS:             3037.03
--------------------------

#### loop4

--------------------------
Durations:       60.09s
Requests:        195679
Avg RT:          152.67ms
Max RT:          585.317ms
Min RT:          59.803ms
Error requests:  0
Valid requests:  195679
QPS:             3056.37
--------------------------

## EpollEventLoop

### loop1

--------------------------
Durations:       60.07s
Requests:        186404
Avg RT:          161.60ms
Max RT:          769.791ms
Min RT:          62.629ms
Error requests:  0
Valid requests:  186404
QPS:             3102.88
--------------------------

### loop2

--------------------------
Durations:       60.04s
Requests:        188606
Avg RT:          158.21ms
Max RT:          402.622ms
Min RT:          56.726ms
Error requests:  0
Valid requests:  188606
QPS:             3141.41
--------------------------

### loop3

--------------------------
Durations:       60.10s
Requests:        189748
Avg RT:          157.53ms
Max RT:          399.215ms
Min RT:          59.836ms
Error requests:  0
Valid requests:  189748
QPS:             3157.08
--------------------------

### loop4

--------------------------
Durations:       60.06s
Requests:        191833
Avg RT:          156.00ms
Max RT:          1033.703ms
Min RT:          60.854ms
Error requests:  0
Valid requests:  191833
QPS:             3194.24
--------------------------

**Comment by zonghaishang** [comment]
@lexburner please provide test code, :-)

**Comment by beiwei30** [comment]
any further update on this pull request @zonghaishang

**Comment by zonghaishang** [comment]
@beiwei30 I will modify it again as suggested by the code review.

**Comment by lexburner** [comment]
Hi
Here is my repository for dubbo epoll  benchmark : https://github.com/lexburner/dubbo-epoll-benchmark <https://github.com/lexburner/dubbo-epoll-benchmark>

Pls add the consumer and provider using your feature/epoll dubbo branch to test it.
